AdminWebsite/AdminWebsite/ClientApp/src/app/work-allocation/work-allocation.component.ts

```diff
@@ -1,18 +1,241 @@
+            `Row ${rowNumber}, Entry ${entryNumber}-${entryNumber + 1} -`
+        );
+
+        return isStartTimeCellValid && isEndTimeCellValid && isStartTimeBeforeEndTime;
+    }
+
+    handleFileInput(file: File) {
+        this.resetErrors();
+
+        if (!file) {
+            return;
+        }
+
+        this.workingHoursFile = file;
+
+        if (this.workingHoursFile.size > this.maxFileUploadSize) {
+            this.isWorkingHoursFileUploadError = true;
+            this.workingHoursFileUploadErrors.push(`File cannot be larger than ${this.maxFileUploadSize / 1000}kb`);
+        }
+    }
+
+    isDelimiterValid(time: string, rowNumber: number, entryNumber: number) {
+        let isValid = true;
+
+        const timeArray = time.split(this.timeDelimiter);
+
+        if (timeArray.length !== 2) {
+            this.workingHoursFileUploadErrors.push(
+                `Row ${rowNumber}, Entry ${entryNumber} - Incorrect delimiter used. Please use a colon to separate the hours and minutes.`
+            );
+            isValid = false;
+        }
+
+        return isValid;
+    }
+
+    isNonWorkingDay(startTime: string, endTime: string) {
+        if (!startTime && (!endTime || endTime === '\r')) {
+            return true;
+        }
+
+        return false;
+    }
+
+    isNonWorkingDayError(startTime: string, endTime: string, errorLocationMessaage: string = '') {
+        if (errorLocationMessaage) {
+            errorLocationMessaage = errorLocationMessaage + ' ';
+        }
+
+        if (!startTime && endTime && endTime !== '\r') {
+            this.workingHoursFileUploadErrors.push(`${errorLocationMessaage}Start time is blank`);
+            return true;
+        }
+
+        if (startTime && (!endTime || endTime === '\r')) {
+            this.workingHoursFileUploadErrors.push(`${errorLocationMessaage}End time is blank`);
+            return true;
+        }
+
+        return false;
+    }
+
+    resetErrors() {
+        this.isWorkingHoursFileUploadError = false;
+        this.workingHoursFileUploadErrors = [];
+    }
+
+    readFile(file: File) {
+        const reader = new FileReader();
+        reader.readAsText(file);
+        reader.onload = e => this.readWorkAvailability(e.target.result as string);
+    }
+
+    readWorkAvailability(text: string): any {
+        const delimiter = ',';
+
+        const userWorkAvailabilityRows = text.split('\n');
+        // Remove headings rows
+        userWorkAvailabilityRows.splice(0, 2);
+
+        const workAvailabilities: WorkAvailability[] = [];
+
+        userWorkAvailabilityRows.forEach((row, index) => {
+            const values = row.split(delimiter);
+
+            const workAvailability = new WorkAvailability();
+            workAvailability.username = values[0];
+
+            const workingHours: DayWorkingHours[] = [];
+
+            let dayOfWeekId = 0;
+            let dayWorkingHours: DayWorkingHours;
+
+            for (let i = 1; i < values.length; i += 2) {
```

Comment:
Out of curiosity, why do we iterate over the columns of the row, in 2s?

Reply:
This is because to consume the data we need to take both the start time and the end time into account at the same time, which represents consuming 2 columns.